compiler/src/dotty/tools/dotc/typer/QuotesAndSplices.scala

```diff
@@ -550,19 +299,43 @@ object QuotesAndSplices {
   /** Context used to type the contents of a quote pattern */
   def quotePatternContext(isTypePattern: Boolean)(using Context): Context =
     quoteContext.fresh.setNewScope
-      .addMode(if isTypePattern then Mode.QuotedTypePattern else Mode.QuotedExprPattern)
-      .retractMode(Mode.Pattern) // TODO do we need Mode.QuotedPattern?
+      .retractMode(Mode.Pattern)
       .setProperty(TypeVariableKey, collection.mutable.Map.empty)
 
   /** Context used to type the contents of a quote pattern splice */
   def quotePatternSpliceContext(using Context): Context =
     spliceContext
       .retractMode(Mode.QuotedPatternBits)
       .addMode(Mode.Pattern)
-      .withOwner(quotePatternOwner(ctx))
+      .withOwner(quotePatternOuterContext(ctx).owner)
 
-  /** First outer context owner that is outside of a quoted pattern context. */
-  private def quotePatternOwner(ctx: Context): Symbol =
-    if ctx.mode.isQuotedPattern then quotePatternOwner(ctx.outer) else ctx.owner
+  /** First outer context that is outside of a quoted pattern. */
+  def quotePatternOuterContext(ctx: Context): Context =
+    if ctx.mode.isQuotedPattern then quotePatternOuterContext(ctx.outer) else ctx
 
+  private[QuotesAndSplices] class TreeMapWithVariance extends TreeMap:
+      private var variance: Int = 1
+
+      def inContravariantPosition: Boolean = variance == -1
+
+      inline private def atVariance[T](v: Int)(op: => T): T = {
+        val saved = variance
+        variance = v
+        val res = op
+        variance = saved
+        res
+      }
+
+      override def transform(tree: Tree)(using Context) = tree match
+        case tree @ AppliedTypeTree(tpt, args) =>
```

Comment:
Maybe that doesn't matter for the transform you're doing, but this isn't complete, it's at least missing TypeBoundsTree, LambdaTypeTree as well as method parameters in DefTrees.

Reply:
It might be a problem, but we must change this behavior in a minor release. This will not affect soundness of the pattern match, but it will change the behavior of existing programs. This is only used to decide which bound of the type we should return. 

This is related to the type inference of type variable bounds and shares some of the same problems and limitations. At some point, we will need a way to allow users to state explicitly if they want to solve for a lower or upper bound. This inference would become redundant or a warning like the one we have for the type bounds.